include/swift/SIL/OwnershipUtils.h

```diff
@@ -200,7 +200,89 @@ bool isOwnershipForwardingInst(SILInstruction *i);
 
 bool isGuaranteedForwardingInst(SILInstruction *i);
 
-struct BorrowScopeIntroducerKind {
+struct BorrowScopeIntroducingOperandKind {
+  using UnderlyingKindTy = std::underlying_type<SILInstructionKind>::type;
+
+  enum Kind : UnderlyingKindTy {
+    BeginBorrow = UnderlyingKindTy(SILInstructionKind::BeginBorrowInst),
+    BeginApply = UnderlyingKindTy(SILInstructionKind::BeginApplyInst),
+  };
+
+  Kind value;
+
+  BorrowScopeIntroducingOperandKind(Kind newValue) : value(newValue) {}
+  BorrowScopeIntroducingOperandKind(
+      const BorrowScopeIntroducingOperandKind &other)
+      : value(other.value) {}
+  operator Kind() const { return value; }
+
+  static Optional<BorrowScopeIntroducingOperandKind>
+  get(SILInstructionKind kind) {
+    switch (kind) {
+    default:
+      return None;
+    case SILInstructionKind::BeginBorrowInst:
+      return BorrowScopeIntroducingOperandKind(BeginBorrow);
+    case SILInstructionKind::BeginApplyInst:
+      return BorrowScopeIntroducingOperandKind(BeginApply);
+    }
+  }
+
+  void print(llvm::raw_ostream &os) const;
+  LLVM_ATTRIBUTE_DEPRECATED(void dump() const, "only for use in the debugger");
+};
+
+/// Operands that create a new borrow scope for an owned value passed as the
+/// operand's value.
```

Comment:
I'm find this a little confusing. When an instruction produces a borrowed value, all of its uses are using the borrowed value. So how is an *operand* introducing the borrow?

Maybe this is just a BorrowedOperand?